### ISSUE TITLE
Working with 1.18 with 2 playwrights when the 2 object gets closed it throws HeapSpace

### ISSUE THREAD

**Issue by cortizcuellar** [microsoft/playwright-java#827]
Here is the info 

https://stackoverflow.com/questions/71242763/java-playwright-throws-java-lang-outofmemoryerror-java-heap-space-closing-a-pla?noredirect=1

And when we connect to a wsEndPoint the problem arises when we try to close Page and Browser objects but not with Playwright object as you can see here.

```
20:39:13.279 [Thread-16] WARN  c.p.h.HeadlessBrowserConfig$DataPerThread.close(219) - Closing [PageImpl] object:[838546035].
java.io.IOException: Stream closed
	at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
	at java.io.OutputStream.write(OutputStream.java:116)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at com.microsoft.playwright.impl.WriterThread.run(PipeTransport.java:167)
java.io.IOException: Stream closed
	at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
	at java.io.OutputStream.write(OutputStream.java:116)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at java.io.FilterOutputStream.close(FilterOutputStream.java:158)
	at com.microsoft.playwright.impl.PipeTransport.close(PipeTransport.java:93)
	at com.microsoft.playwright.impl.PipeTransport.poll(PipeTransport.java:71)
	at com.microsoft.playwright.impl.Connection.processOneMessage(Connection.java:159)
	at com.microsoft.playwright.impl.ChannelOwner.runUntil(ChannelOwner.java:101)
	at com.microsoft.playwright.impl.Connection.sendMessage(Connection.java:107)
	at com.microsoft.playwright.impl.ChannelOwner.sendMessage(ChannelOwner.java:94)
	at com.microsoft.playwright.impl.JsonPipe.send(JsonPipe.java:46)
	at com.microsoft.playwright.impl.Connection.internalSendMessage(Connection.java:135)
	at com.microsoft.playwright.impl.Connection.sendMessageAsync(Connection.java:111)
	at com.microsoft.playwright.impl.Connection.sendMessage(Connection.java:107)
	at com.microsoft.playwright.impl.ChannelOwner.sendMessage(ChannelOwner.java:94)
	at com.microsoft.playwright.impl.PageImpl.close(PageImpl.java:547)
	at com.microsoft.playwright.Page.close(Page.java:3237)
20:39:13.321 [Thread-16] WARN  c.p.h.HeadlessBrowserConfig$DataPerThread.close(223) - Exception [Failed to read message from driver, pipe closed.]Closing [PageImpl] object:[838546035].

```

**Comment by yury-s** [comment]
You are running out of memory, looks like the machine doesn't have enough memory to run multiple playwrights on it. If you still think this is a bug in the playwright please provide a repro that we could run locally.

**Comment by cortizcuellar** [comment]
I am running with **-Xms3g -Xmx3g**  when the shutdown is call i am using like 800 mb later the Out of memory exception is thrown when i call **shutdown** on the second playwright i will test this the same scenario with more memory.

**Comment by cortizcuellar** [comment]
Running with **-Xms20g -Xmx20g**  with connecting to a _**wsEndPoint Browserless**_ in this case. When trying to close **Page** object it throws 
```
19:15:42.135 [Thread-16] DEBUG c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(220) - Page [2117799917] is closed [false] browser is connected:[true].
19:15:42.136 [Thread-16] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(238) - Closing [PageImpl] object:[2117799917].
java.io.IOException: Stream closed
	at java.lang.ProcessBuilder$NullOutputStream.write(ProcessBuilder.java:433)
	at java.io.OutputStream.write(OutputStream.java:116)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	at com.microsoft.playwright.impl.WriterThread.run(PipeTransport.java:167)

```
_20 GB is a lot of memory._
**_Any clue._**

**Comment by yury-s** [comment]
Can you share a repro that we could run locally?

**Comment by cortizcuellar** [comment]
> Can you share a repro that we could run locally?

I will try that today mate. :)

**Comment by cortizcuellar** [comment]
When running the process with **20G of ram memory** the 2 page object it hangs the process as you can see here it takes more than 10 minutes hang i have try even longer 2 or 3 hours and continues hanged.
You can see in the top of the image the arguments at Java process. 
![image](https://user-images.githubusercontent.com/85301269/156566772-d94bcf45-d439-40f4-be6e-684f5b8f19af.png)

```
08:23:03.458 [Thread-13] WARN  c.p.w.ProxyConfigServiceImpl.shutdown(223) - HeadlessBrowserConfig:[1092398950]
08:23:03.462 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig.shutdown(388) - Closing resources[Start] in headlessBrowser [1092398950] object
08:23:03.462 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(219) - Shutdown process start at [269559688] with thread:[Thread-13].
08:23:03.463 [Thread-13] DEBUG c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(222) - Page [1641736644] is closed [false] browser is connected:[true].
08:23:03.463 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(240) - Closing [PageImpl] object:[1641736644] with free memory:[16071.636199951172].
08:23:03.474 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(244) - Closed [PageImpl] object:[1641736644]. Free---->16071.636199951172 [MB] took:[6] ms.
08:23:03.474 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(240) - Closing [BrowserImpl] object:[184488816] with free memory:[16071.636199951172].
08:23:03.474 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(244) - Closed [BrowserImpl] object:[184488816]. Free---->16071.636199951172 [MB] took:[0] ms.
08:23:03.474 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(240) - Closing [PlaywrightImpl] object:[1288055767] with free memory:[16071.636199951172].
08:23:03.713 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(244) - Closed [PlaywrightImpl] object:[1288055767]. Free---->15754.336112976074 [MB] took:[233] ms.
08:23:03.713 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(230) - Shutdown process end at [269559688] with thread:[Thread-13].
08:23:03.722 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(219) - Shutdown process start at [993683470] with thread:[Thread-13].
08:23:03.722 [Thread-13] DEBUG c.p.h.HeadlessBrowserConfig$DataPerThread.shutdown(222) - Page [1522931126] is closed [false] browser is connected:[true].
08:23:03.723 [Thread-13] INFO  c.p.h.HeadlessBrowserConfig$DataPerThread.close(240) - Closing [PageImpl] object:[1522931126] with free memory:[15754.336112976074].

As you can see the times of closing the 1 bunch of objects is just like **0 ms or 6 ms** the problem arises when we try the close the  objects created at 2 thread

Thanks a lot and God bless.


```

**Comment by cortizcuellar** [comment]
If i run with flag **-XX:+PrintGCDetails** i can the next and immediately after the call to the 2 Page object shutdown method the message 

`[GC (Allocation Failure) [PSYoungGen: 4633312K->67619K(6116864K)] 4682338K->1909903K(20098048K), 0.6676778 secs] [Times: user=1.06 sys=0.36, real=0.67 secs]`

![image](https://user-images.githubusercontent.com/85301269/156570240-a690ff85-3796-4292-804b-2191a26505e9.png)

**Comment by cortizcuellar** [comment]
Strangely inside IntellijIdea applying **System.exit(0)** with **10 playwrights** with -**Xmx:5g** works like a charm. I will dive more on this.

![image](https://user-images.githubusercontent.com/85301269/156840736-36f9bd68-e90f-463d-a635-4f8fd870c3f2.png)

**Comment by cortizcuellar** [comment]
I can reproduce with this sample of course the running code is harder and better designed but i think will reproduce it..

```
import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class StreamClosed{

    private final List<DataPerThread>dataPerThreads = new ArrayList<>();

    private final CountDownLatch latch = new CountDownLatch(5);

    private static final double MB_VALUE = 1024 * 1024;

    private final List<String> urls = new ArrayList<>();//will be using only one in all 5 threads for the sake of simplicity

	interface ShutdownEvent {
		void shutdown();
	}

    public static void main(String[] args) throws Exception{
        final StreamClosed clazz = new StreamClosed();
        final Runnable hook = ()->clazz.dataPerThreads.forEach(DataPerThread::shutdown);//CTRL + C WAS PRESSED
        Runtime.getRuntime().addShutdownHook(new Thread(hook));
        ExecutorService service = Executors.newFixedThreadPool(5);
        for(final String url:clazz.urls){
            final ThreadCaller thread = clazz.new ThreadCaller("http://myexternalip.com/raw");
            System.out.println("Send work with thread: "+thread.getName());
            service.execute(thread);
        }
        clazz.waitForChildThreadsToComplete();
        service.shutdown();
        service.shutdownNow();
        System.out.println("Ends....");
        try {
            Thread.sleep(240_000);
            clazz.dataPerThreads.forEach(DataPerThread::shutdown);//IF HERE NOT CTRL+C WAS PRESSED
        }catch (final Exception e){
            e.printStackTrace();
        }
    }

    protected void waitForChildThreadsToComplete() {
        try {
            latch.await();
            System.out.println("All child threads have completed.");
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private final class ThreadCaller extends Thread{
        private final String url;

        public ThreadCaller(final String url) {
            this.url = url;
        }

        @Override
        public void run() {
            new DataPerThread(url).run();
        }
    }

    private final class DataPerThread implements ShutdownEvent,Runnable {
        protected final Logger LOGGER = LoggerFactory.getLogger(DataPerThread.class);

        private final Playwright playwright;

        private final String url;
        private final BrowserType browserType;
        private Browser browser;
        private Page page;
        private final int playwrightHashCode;

        public DataPerThread(final String url) {
            LOGGER.info("Building DataPerThreadObject for Thread:[{}]",Thread.currentThread().getName());
            LOGGER.info("Creating playwright with thread [{}].",Thread.currentThread().getName());
            final long startPlayWright = System.currentTimeMillis();
            this.url = url;
            this.playwright = Playwright.create();
            dataPerThreads.add(this);
            this.playwrightHashCode = System.identityHashCode(this.playwright);
            LOGGER.info("Working with Playwright:[{}] took:[{}] ms.",this.playwrightHashCode,(System.currentTimeMillis()-startPlayWright));
            LOGGER.debug("Creating browserType [{}]",this.playwrightHashCode);
            final long creatingBrowsersInfo = System.currentTimeMillis();
            this.browserType = playwright.chromium();
            LOGGER.debug("Created browserType [{}] with Playwright:[{}]",System.identityHashCode(this.browserType),this.playwrightHashCode);
            fillPlayWrightProxy();
            LOGGER.info("Working with browsers objects :[{}] with Playwright [{}] took:[{}] ms.",System.identityHashCode(browserType),playwrightHashCode,(System.currentTimeMillis()-creatingBrowsersInfo));
            LOGGER.info("Builded DataPerThreadObject [{}] for Thread:[{}]",System.identityHashCode(this),Thread.currentThread().getName());
        }

        @Override
        public void run() {
            System.out.println("Connecting to: "+url+" with Thread:[{}]"+Thread.currentThread().getName());
            try {
                System.out.println("Requesting: "+url+" with page: "+(System.identityHashCode(page)));
                System.out.println(page.navigate(url).body().length);
            }catch(final Exception e){
                System.out.println(e.getLocalizedMessage());
            }
            latch.countDown();
            System.out.println("Left: "+latch.getCount());
        }

        private void fillPlayWrightProxy(){
            LOGGER.info("BrowserType: [{}]",System.identityHashCode(browserType));
            this.browser = browserType.connect("wss://chrome.browserless.io/playwright?token="+token);
            LOGGER.debug("Created pageProvider [{}] with BrowserType:[{}]",System.identityHashCode(this.browser),System.identityHashCode(this.browserType));
            this.page = newPage();
        }

        private Page newPage(){
            LOGGER.debug("Trying to open page with Playwright Object:[{}]",playwrightHashCode);
            final long start = System.currentTimeMillis();
            final Page localPage = this.browser.newPage();
            localPage.onClose(this::onClosePage);
            final long end = System.currentTimeMillis();
            LOGGER.debug("Building page [{}] for thread [{}] with playwright: [{}] took {} ms.",System.identityHashCode(localPage),Thread.currentThread().getName(),playwrightHashCode,(end - start));
            return localPage;
        }

        private void onClosePage(final Page page){
            LOGGER.debug("Page [{}] will be closed shutDownProcess started:[{}].",System.identityHashCode(page),false);
        }

        @Override
        public void shutdown() {
            final String threadName = Thread.currentThread().getName();
            final long hashcode = System.identityHashCode(this);
            LOGGER.info("Shutdown process start at [{}] with thread:[{}].",hashcode,threadName);
            final boolean pageIsClosed = page.isClosed();
            final boolean browserIsConnected = browser.isConnected();
            LOGGER.debug("Page [{}] is closed [{}] browser is connected:[{}] browserContext:[{}].",System.identityHashCode(page),pageIsClosed,browserIsConnected,System.identityHashCode(page.context()));
            if(!pageIsClosed && browserIsConnected){
                close(page);
            }
            if(browserIsConnected){
                close(browser);
            }
            close(playwright);
            LOGGER.info("Shutdown process end at [{}] with thread:[{}].",hashcode,threadName);
        }

        private double getFreeMemory(){
            return Runtime.getRuntime().freeMemory() / MB_VALUE;
        }

        private DataPerThread close(final AutoCloseable closeable){
            if(closeable == null){
                return this;
            }
            final String className = closeable.getClass().getSimpleName();
            final int hashcode = closeable.hashCode();
            try{
                LOGGER.info("Closing [{}] object:[{}] with free memory:[{}] MB.",className,hashcode,getFreeMemory());
                final long startTime = System.currentTimeMillis();
                closeable.close();
                final long endTime = System.currentTimeMillis();
                LOGGER.info("Closed [{}] object:[{}]. Free---->{} [MB] took:[{}] ms.",className,hashcode,getFreeMemory(),(endTime-startTime));
            }catch(final Exception e){
                LOGGER.warn("Exception [{}]Closing [{}] object:[{}].",e.getLocalizedMessage(),className,hashcode);
            }
            return this;
        }
    }
}
```
With 3GB and pressing [Ctrl+C] in sleep it throws 
![image](https://user-images.githubusercontent.com/85301269/158452777-a79c8677-7459-41c8-8f4b-0c4cd0b1b7c5.png)

With 35 GB it hangs forever and [Ctrl+C].

![image](https://user-images.githubusercontent.com/85301269/158453125-304d1a35-bcf8-4d8c-9491-a29dff08bb8e.png)

**Comment by yury-s** [comment]
Does it reproduce with local browsers instead of browserless ?

**Comment by cortizcuellar** [comment]
> Does it reproduce with local browsers instead of browserless ?

I will let you know tomorrow :)

**Comment by cortizcuellar** [comment]
In the previous snippet i was using **browserless** not using **Browserless** i mean instead of using

**this.browser = browserType.connect("wss://chrome.browserless.io/playwright?token="+token);**

Using

**this.browser = browserType.launch();**

With 3GB of ram it closes Page + Browser object but it throws heap closing the 1 playwright object.

![image](https://user-images.githubusercontent.com/85301269/158809527-79a4c64e-d72a-4d4f-a342-a3894256acf0.png)

Trying to understand the dump i see something like this.

![image](https://user-images.githubusercontent.com/85301269/158810190-81bc974a-d6f6-4d98-b6ea-48ab658634d0.png)
![image](https://user-images.githubusercontent.com/85301269/158810441-e5bc7b69-4800-4e60-9a87-840a90e1199f.png)

With 35 GB it hangs closing the page of the second thread.

![image](https://user-images.githubusercontent.com/85301269/158814112-74b1ecd8-92a9-408b-9ed0-f3ddd9588989.png)

Please note that when the shutdown process started is the **Thread-0** is the thread i think that will kill Java process.

As you can see in the code i put  a **.sleep** this is mainly for see what happens when a **Ctrl+C** is call or if not how behave the code when a the regular flow is not interrupted. 

If not **CTRL+C** is not pressed and the code flows normally even with 3GB of ram it closes Smoothly. As you can see here the thread is **main** instead of **Thread-0**

![image](https://user-images.githubusercontent.com/85301269/158815865-58f92975-456c-4497-b0fc-afa755e3cc6c.png)

It seems it have what to do which **Thread** closes the objects :(

**Comment by yury-s** [comment]
Try running it with `BROWSER=pw:channel` (and `DEBUG=pw:browser` if that doesn't give enough info) and see what messages are flowing from the driver to the client. My hunch is that the close operation may take longer (especially on Windows) because we try to delete some temp files and that takes a while. While this is happening, the playwright client may keep receiving events from the driver without being able to process them.

**Comment by cortizcuellar** [comment]
> Try running it with `BROWSER=pw:channel` (and `DEBUG=pw:browser` if that doesn't give enough info) and see what messages are flowing from the driver to the client. My hunch is that the close operation may take longer (especially on Windows) because we try to delete some temp files and that takes a while. While this is happening, the playwright client may keep receiving events from the driver without being able to process them.

Are this system properties o maven commands? we are using gradle..

**Comment by yury-s** [comment]
> Are this system properties o maven commands? we are using gradle..

These are environment variables.

**Comment by cortizcuellar** [comment]
> > Are this system properties o maven commands? we are using gradle..
> 
> These are environment variables.

I will do this mate and i will let you my kid is in the hospital :(

**Comment by cortizcuellar** [comment]
With 3GB ram **_OutOfMemoryException_**
![image](https://user-images.githubusercontent.com/85301269/159700766-9612288a-8916-4994-92a3-68d38e074d28.png)
With 10GB RAM **_OutOfMemoryException_**
![image](https://user-images.githubusercontent.com/85301269/159702127-5193914d-05f9-4552-a330-8d2b770ebb66.png)
With 15GB ram it hangs closing the page of the second thread.
![image](https://user-images.githubusercontent.com/85301269/159707043-6e4384cf-f019-484e-8905-6ea48901b40e.png)

**Comment by cortizcuellar** [comment]
With Environment variables set.

**Comment by mxschmitt** [comment]
We fixed a related bug in 1.22, could you try to reproduce it there?

**Comment by cortizcuellar** [comment]
Hey mate with the latest Playwright version it works. 1.25 :)
![image](https://user-images.githubusercontent.com/85301269/188504973-75292971-4e9e-4b37-a8d4-fa0286bd4564.png)
